remembering show-hide viz control

moontrip opened the pull request:
This will address https://github.com/VEuPathDB/web-monorepo/issues/935. Currently, only the first part, remembering show/hide viz control, is done. Will work for the second part, tech debt

Initially, I tried to set showVizControl and change prop names accordingly. But while doing that, I realized that there were lots of changes involved, including all Vizs, so I stopped it and named it as hideVizControl instead. @bobular you think it should be?

moontrip commented:
I did also try to use `hideVizControl` under `visualizationPanelConfig `(i.e., at `PanelConfig` and also under `markerConfigurations`). However, I could not resolve one issue, a conflict with `setActiveVisualizationId` when creating New Plot. I think that by doing so, both `setActiveVisualizationId` and `setHideVizInputsAndControls` (basically it is `setHideVizControl`) utilize `updateMarkerConfigurations` callback, thus it seems like there occurs a race condition between the two at `onVisualizationCreated()`, which could not be resolved. Thus, I discarded the idea, and set the `hideVizControl` under appState directly. 

As for the tech debt, it seems to me that there would involve lots of works when removing `isVisible`, thus I set a conditional rendering of `DraggableVisualization` instead

bobular reviewed (commented):
I'm replying to your discussion comment separately. These are just two small observations on the code so far.

bobular commented:
> I did also try to use `hideVizControl` under `visualizationPanelConfig `(i.e., at `PanelConfig` and also under `markerConfigurations`). However, I could not resolve one issue, a conflict with `setActiveVisualizationId` when creating New Plot. I think that by doing so, both `setActiveVisualizationId` and `setHideVizInputsAndControls` (basically it is `setHideVizControl`) utilize `updateMarkerConfigurations` callback, thus it seems like there occurs a race condition between the two at `onVisualizationCreated()`, which could not be resolved. Thus, I discarded the idea, and set the `hideVizControl` under appState directly.

To prepare for future enhancements it would be good for the show/hide status of the vizInputsAndControls to "belong" to the `visualizationPanelConfig`. So the type definition would be something like this in appState.ts:

```
// this bit needs to be io-ts-ified - making sure the new prop is `partial`/optional
type VisualizationPanelConfig = PanelConfig & {
  hideInputsAndControls?: boolean; // for backwards compatibility, default will be falsy
};

export const MarkerConfiguration = t.intersection([
  t.type({
    type: MarkerType,
    selectedVariable: VariableDescriptor,
  }),
  t.partial({
    activeVisualizationId: t.string,
  }),
  t.union([
    t.intersection([
      t.type({
        type: t.literal('barplot'),
        selectedValues: SelectedValues,
        selectedPlotMode: t.union([
          t.literal('count'),
          t.literal('proportion'),
        ]),
        binningMethod: BinningMethod,
        dependentAxisLogScale: t.boolean,
        selectedCountsOption: SelectedCountsOption,
        legendPanelConfig: PanelPositionConfig,
        visualizationPanelConfig: VisualizationPanelConfig,  // <<<< new <<<<
      }),
      // continues...
```
The `hideInputsAndControls` prop passed to `DraggableVisualization` would be
`visualizationPanelConfig.hideInputsAndControls ?? false`

The `setHideInputsAndControls` handler passed to `DraggableVisualization` would be similar to `updateVisualizationPosition` (and all these handlers could be DRYed up by providing them from a hook)


The `onVisualizationCreated()` race condition issue can likely be solved by adding an optional second argument to `setActiveVisualizationId`

```
  const setActiveVisualizationId = useCallback(
    (activeVisualizationId?: string, isNew?: boolean) => {
      if (configuration == null) return;
      updateConfiguration({
        ...configuration,
        activeVisualizationId,
        visualizationPanelConfig: {
          ...visualizationPanelConfig,
          isVisible: !!activeVisualizationId,
          ...(isNew ? { hideInputsAndControls: false } : { }),
        },
      });
    },
    [configuration, updateConfiguration, visualizationPanelConfig]
  );
```
(And we should be able to DRY that up too.)

And then using that only in `onVisualizationCreated`:

```
  function onVisualizationCreated(visualizationId: string) {
    setIsVizSelectorVisible(false);
    setActiveVisualizationId(visualizationId, true);
    setIsSidePanelExpanded(false);
  }
```

Hope that makes sense! Let me know if not :-)

moontrip commented on packages/libs/eda/src/lib/map/analysis/appState.ts:
```diff
@@ -152,6 +152,7 @@ export const AppState = t.intersection([
     activeMarkerConfigurationType: MarkerType,
     markerConfigurations: t.array(MarkerConfiguration),
     isSidePanelExpanded: t.boolean,
+    hideVizControl: t.boolean,
```
@bobular Thank you for your comments 👍 Yes it should be partial as you stated: I missed the case for pre-existing ones.

moontrip commented on packages/libs/eda/src/lib/map/analysis/mapTypes/plugins/BubbleMarkerMapType.tsx:
```diff
@@ -476,26 +476,28 @@ function BubbleLegendsAndFloater(props: MapTypeMapLayerProps) {
-        dimensions={visualizationPanelConfig.dimensions}
-        onPanelDismiss={onPanelDismiss}
-      />
+      {visualizationPanelConfig?.isVisible && (
```
@bobular I thought that one of your suggestions was to set conditional rendering that wraps `DraggableVisualization` like `SubStudies`. I might have misunderstood your intention. So, no I didn't fully try to remove conditional rendering inside `DraggableVisualization`: I briefly tried to remove it but it caused several type errors.

moontrip commented:
@bobular Thank you for your detailed suggestion above 👍 I didn't try to handle both at the `setActiveVisualizationId()` primarily because it might cause some confusion in between `setActiveVisualizationId()` and `setHideVizInputsAndControls()`. Will check if it resolves the issue I had before.

moontrip commented:
@bobular As suggested in your last comment, I refactored codes and removed unnecessary parts in the following: https://github.com/VEuPathDB/web-monorepo/pull/967/commits/259d4a99085e075b09af8031a71d5989eb4d2182 Overall, it seems to work fine

bobular reviewed (commented):
This is great - I'm glad it worked :relieved: 

I've pushed a commit that moves all the show/hide logic out of MapAnalysis (before Dave sees it!)

moontrip commented:
> This is looking good.
> 
> I have one request before you merge:
> 
> Please could you either resolve the issue of the duplicated conditional rendering of the floating viz panel or make a new tech debt ticket describing it?
> 
> Thanks!

Thank you for your reviews and comments, @bobular. I will make a new tech debt ticket about the conditional rendering as it is beyond the show/hide control stuff.

moontrip commented:
Tech debt ticket is separated into https://github.com/VEuPathDB/web-monorepo/issues/1008
